File: javascript/smallest_distance_between_two_array.js
function binarySearch(arr, number) {
    h = arr.length - 1;
    l = 0;
    let mid;
    let smallest_distance = Infinity;
    let result;
    while(l<=h) {
        mid = Math.round((h+l)/2);
        let curr_distance = Math.abs(number - arr[mid]);
        if(curr_distance < smallest_distance) {
            smallest_distance = curr_distance;
            result = mid;
        }
        if(arr[mid] === number) {
            break;
        }
        else if(arr[mid]>number) {
            h = mid - 1;
        }
        else if(arr[mid]<number) {
            l = mid + 1;
        }
    }
    return result;
}
arr = [-2,-4,0,3,2,5,6];
arr_distance = [-4,-6,4,9];
arr.sort();
let distance = Infinity;
let value1, value2;
for(let i = 0; i<arr_distance.length;i++) {
    let result_idx = binarySearch(arr, arr_distance[i]);

    let cur_distance =  Math.abs(arr[result_idx] -  arr_distance[i]);
    if(cur_distance < distance) {
        distance = cur_distance;
        value1 = arr[result_idx];
        value2 = arr_distance[i];
    }
}
console.log(value1);
console.log(value2);

File: dart/smallest_distance_between_two_array.dart
import 'binary_search.dart';

double smallestDistanceBetweenTwoArray(
    List<double> array1, List<double> array2) {
  // smallest distance
  double smallest_distance = double.infinity;
  if (array1.length <= array2.length) {
    for (int i = 0; i < array1.length; i++) {
      int index = binarySearchNearestIndex(array2, array1[i]);
      if (smallest_distance > (array1[i] - array2[index]).abs()) {
        smallest_distance = (array1[i] - array2[index]).abs();
      }
    }
  } else {
    for (int i = 0; i < array2.length; i++) {
      int index = binarySearchNearestIndex(array1, array2[i]);
      if (smallest_distance > (array2[i] - array1[index]).abs()) {
        smallest_distance = (array2[i] - array1[index]).abs();
      }
    }
  }
  return smallest_distance;
}
